crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -0,0 +1,123 @@
+            // information regarding third-party dependencies. That means that we don't need
+            // to care about any possible `lib64/site-packages` directories, since
+            // [the `sys`-module documentation] states that `sys.platlibdir` is *only* ever
+            // used for C extensions, never for pure-Python modules.
+            //
+            // [the non-Windows branch]: https://github.com/python/cpython/blob/a8be8fc6c4682089be45a87bd5ee1f686040116c/Lib/site.py#L401-L410
+            // [the `sys`-module documentation]: https://docs.python.org/3/library/sys.html#sys.platlibdir
+            Self::Unix => system
+                .read_directory(&venv_path.join("lib"))?
+                .flatten()
+                .find_map(|entry| {
+                    if !entry.file_type().is_directory() {
+                        return None;
+                    }
+                    let path = entry.path();
+
+                    // The `python3.x` part of the `site-packages` path can't be computed from
+                    // the `--target-version` the user has passed, as they might be running Python 3.12 locally
+                    // even if they've requested that we type check their code "as if" they're running 3.8.
+                    //
+                    // The `python3.x` part of the `site-packages` path *could* be computed
+                    // by parsing the virtual environment's `pyvenv.cfg` file.
+                    // Right now that seems like overkill, but in the future we may need to parse
+                    // the `pyvenv.cfg` file anyway, in which case we could switch to that method
+                    // rather than iterating through the whole directory until we find
+                    // an entry where the last component of the path starts with `python3.`
+                    path.components()
+                        .next_back()
+                        .is_some_and(|last_part| last_part.as_str().starts_with("python3."))
+                        .then_some(path.join("site-packages"))
+                })
+                .ok_or(SitePackagesDiscoveryError::NoSitePackagesDirFound),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SitePackagesDiscoveryError {
+    #[error("Failed to search the virtual environment directory for `site-packages` due to {0}")]
+    CouldNotReadLibDirectory(#[from] io::Error),
+    #[error("Could not find the `site-packages` directory in the virtual environment")]
+    NoSitePackagesDirFound,
+}
+
+/// Given a validated, canonicalized path to virtual environment,
+/// return a list of `site-packages` directories that are available from that environment.
+///
+/// TODO: Currently we only ever return 1 path from this function:
+/// the `site-packages` directory that is actually inside the virtual environment.
+/// Some `site-packages` directories are able to also access system `site-packages` and
+/// user `site-packages`, however.
```

Comment:
Planning on tackling this TODO in a followup. I wanted to keep the logic in this PR as simple as possible for now.